File: matlab/KalmanBase.m
classdef (Abstract) KalmanBase < handle
    % KalmanBase  Base class for Kalman filters and smoothers implemented 
    %             in Matlab (there are separate interface classes for the
    %             native and Java implementations).
    %
    % KalmanBase fields (properties):
    %    steps    - A cell array of structures that represent the time steps
    %    current  - Represents the current time step
    %    options  - Structure containing options
    %  
    % KalmanBase abstract methods:
    %    evolve   - Evolve the state using a linear matrix equation 
    %    observe  - Provide observations of the current state 
    %    smooth   - Compute smooth estimates of all the stored states
    % 
    % KalmanBase methods
    %    estimate - Return the most up to date estimate of a state vector
    %               and its covariance matrix
    %    forget   - Forget the oldest steps to save memory meory 
    %    rollback - Roll back the object to an earlier step 
    %    latest   - The index of the last step that was observed 
    %    earliest - The index of the earliest step that has not been forgoten
    %    gather   - return all the estimates concatenated into a column
    %               vector
    %    pertest  - measure the performance of Kalman filtering or
    %               smoothing
    %
    % Subclasses must define a rollbackStepWithIndex method that should reconstruct
    % the current structure to the point just after the call to evolve() on
    % a given step in the sequence.
    %
    % Copyright 2024 Sivan Toledo, Tel Aviv University.

    properties (Access = protected)
        steps;   % must have at least a member 'step'
        current; % the current state; moved to steps at the end of observe
        options;
    end

    methods (Abstract, Access = public)
        evolve(kalman,n_i,H_i,F_i,c_i,K_i)
        observe(kalman,G_i,o_i,C_i)
        smooth(kalman)
    end

    methods (Abstract, Access = protected)
        rollbackStepWithIndex(kalman,i)
    end

    methods (Access = protected)
        function i = indexOfStep(kalman,s)
            if isempty(kalman.steps)
                error('invalid internal state');
            end

            l = length(kalman.steps);
            earliest = kalman.steps{ 1 }.step;
            i = s - earliest + 1;
            if i > l || i < 1
                error('invalid internal state');
            end
        end
    end


    methods (Access = public)
        function kalman = KalmanBase(options)
            kalman = kalman@handle();
            kalman.steps = {};
            kalman.options = options;
        end

        % no need for a destructor

        function i = earliest(kalman)
            %EARLIST   The index of the oldest step that has not been
            %          forgotten
            %
            %   kalman.EARLIEST() returns the index of the earliest step
            %   that has not been forgoten. Step numbers start from zero.
            if isempty(kalman.steps)
                i = -1;
            else
                i = kalman.steps{ 1 }.step;
            end
        end

        function i = latest(kalman)
            %LATEST   The index of the last step that was observed.
            %
            %   kalman.LATEST() returns the index of the latest step that 
            %   was observed.
            if isempty(kalman.steps)
                i = -1;
            else
                i = kalman.steps{ length(kalman.steps) }.step;
            end
        end

        function forget(kalman,s)
            %FORGET  Forget the oldest steps to save memory meory.
            %   
            %   kalman.FORGET(s) forgets steps s and older. 
            %
            %   kalman.FORGET()  forgets all but the last step.
            
            %l = length(kalman.steps);
            % earliest = kalman.steps{ 1 }.step;
            % if nargin<2
            %     ptr_s = l-1;
            % else
            %     ptr_s = s - earliest + 1;
            % end
            l = length(kalman.steps);
            ptr_s = kalman.indexOfStep(s);
            kalman.steps = { kalman.steps{ptr_s+1:l} };
        end

        function rollback(kalman,s)
            %ROLLBACK  Rolls the object back to just after the evolution
            %          to step s (but before the observation in step s).
            %
            %   kalman.ROLLBACK(s) rolls back the filter. Steps later than
            %   s are discareded, and so is the observation of step s, if
            %   there was any.
            % 
            %   This method allows you to roll back the object after it was
            %   used for prediction of future state that have not been
            %   observed yet. Typically steps s+1 and higher do not have
            %   any observation at the time of the rollback.
            
            % l = length(kalman.steps);
            % earliest = kalman.steps{ 1 }.step;
            % ptr_s = s - earliest + 1;
            % if ptr_s > l || ptr_s < 1
            %     warning('cannot roll back to this state (too old or future');
            % else
            ptr_s = kalman.indexOfStep(s);
                kalman.rollbackStepWithIndex(ptr_s); % abstract method to return this step to just after evolve
                kalman.steps = { kalman.steps{1:ptr_s-1} }; % this deletes the step we rolled back to!
            % end
        end

        function [estimate,cov] = estimate(kalman,s)
            %ESTIMATE  Obtain the most recent estimate of the state of a step
            %          and the covariance of the estimate.
            %
            %   [estimate,cov] = kalman.ESTIMATE(s) returns an estimate of 
            %   the state of step s, which must still be in memory. It also 
            %   returns the covariance of the estimate.
            %
            %   [estimate,cov] = kalman.ESTIMATE() returns an estimate of
            %   the state of the latest step.
            %
            %   If kalman.smooth() was not called after step s was
            %   observed, the estimate is a filtered estimate. Otherwise it
            %   is the most recent smoothed estimate.
            
            if nargin<2
                s = kalman.latest();
            end
            
            ptr_s = kalman.indexOfStep(s);

            if isfield(kalman.steps{ptr_s},'estimatedState') && ~isempty(kalman.steps{ptr_s}.estimatedState)
                estimate = kalman.steps{ptr_s}.estimatedState;
                cov      = kalman.steps{ptr_s}.estimatedCovariance;
            else
                estimate = NaN * zeros(kalman.steps{ptr_s}.dimension,1);
                cov      = CovarianceMatrix(NaN*eye(kalman.steps{ptr_s}.dimension),'W');
                warning('state is currently underdetermined');
            end
        end

        function u = gather(kalman)
            l = length(kalman.steps);
            N = 0;
            for i=1:l
                N = N + kalman.steps{i}.dimension;
                 %[range' u(range)]
            end
            %N = kalman.steps{l}.estimateStart + kalman.steps{l}.dimension - 1;
            u = zeros(N,1);

            j = 1;
            for i=1:l
                range = j:j + kalman.steps{i}.dimension - 1;
                u(range) = kalman.estimate(kalman.steps{i}.step);
                %[range' u(range)]
                j = j + kalman.steps{i}.dimension;
                
            end
        end
        
        function t = perftest(kalman,H,F,c,K,G,o,C,count,decimation)
            % Stress testing
            n = size(G,2);
            m = size(G,1);

            %count
            %decimation
            %floor(count/decimation)

            t = zeros(floor(count/decimation),1);

            tstart = tic;
            
            j = 1;

            for i=1:count
                evolve(kalman,n,H,F,c,K);
                observe(kalman,G,o,C);
                [e,cov] = estimate(kalman);
                forget(kalman);
                if rem(i,decimation)==0
                    telapsed = toc(tstart);
                    t(j) = telapsed/decimation;
                    j = j+1;
                    tstart = tic;
                end
            end
        end

        function t = perftestSmoothing(kalman,H,F,c,K,G,o,C,count)
            % Stress testing
            n = size(G,2);

            t = NaN * zeros(2,1);

            tstart = tic;
            
            for i=1:count
                evolve(kalman,n,H,F,c,K);
                observe(kalman,G,o,C);
            end
            t(2,1) = toc(tstart);
            smooth(kalman);

            t(1,1) = toc(tstart);
        end

        
    end % methods
end

File: python/kalman_base.py
import numpy as np
from covariance_matrix import CovarianceMatrix
from abc import ABC, abstractmethod
from collections import OrderedDict
from flexible_array import FlexibleArray  # Importing the previously created FlexibleArray

class KalmanBase(ABC):
    """
    KalmanBase: Abstract base class for Kalman filters and smoothers.
    
    Attributes:
        steps (FlexibleArray): Stores time step structures.
        current (dict): Represents the current time step.
        options (dict): Stores configuration options.
    """

    def __init__(self, options):
        """
        Initialize the KalmanBase object.
        
        Args:
            options (dict): Options for the Kalman filter.
        """
        self.steps = FlexibleArray()  # Using FlexibleArray to store steps
        self.current = None  # Current state, moved to steps at the end of observe
        self.options = options

    @abstractmethod
    def evolve(self, n_i, H_i, F_i, c_i, K_i):
        """Evolve the state using a linear matrix equation."""
        pass

    @abstractmethod
    def observe(self, G_i, o_i, C_i):
        """Provide observations of the current state."""
        pass

    @abstractmethod
    def smooth(self):
        """Compute smooth estimates of all stored states."""
        pass

    @abstractmethod
    def rollback_step_with_index(self, i):
        """Reconstructs the object to the state just after evolve() at step i."""
        pass

    def index_of_step(self, s):
        """
        Get the index of step s in the FlexibleArray.

        Args:
            s (int): Step number.

        Returns:
            int: Index in steps.
        
        Raises:
            ValueError: If step is not found or state is invalid.
        """
        #if self.steps.size() == 0:
        #    raise ValueError("Invalid internal state: No steps stored.")

        #earliest = self.steps.first()["step"]
        #i = s - earliest
        #print('xxx')
        #print(f">>> {i} {s} {earliest} {self.steps.size()}")
        #print( self.steps.data.keys() )

        #if i < 0 or i >= self.steps.size():
        #    raise ValueError("Invalid internal state: Step out of range.")

        #return i
        # trivial in this implementation
        return s

    def earliest(self):
        """
        Returns the index of the earliest step that has not been forgotten.

        Returns:
            int: Index of the earliest step, or -1 if no steps exist.
        """
        return self.steps.first()["step"] if self.steps.size() > 0 else -1

    def latest(self):
        """
        Returns the index of the latest step that was observed.

        Returns:
            int: Index of the latest step, or -1 if no steps exist.
        """
        return self.steps.last()["step"] if self.steps.size() > 0 else -1

    def forget(self, s=None):
        """
        Forget the oldest steps to save memory.

        Args:
            s (int, optional): Forget all steps <= s. Defaults to keeping only the last step.
        """
        if self.steps.size() == 0:
            return

        if s is None:
            s=self.latest() - 1
        
        while self.earliest() != -1 and self.earliest() <= s:
            #print(f"rollback calling drop_last s={s} earliest={self.earliest()}")
            self.steps.drop_first()

    def rollback(self, s):
        """
        Rollback the object to just after the evolution of step s (but before observation of step s).

        Args:
            s (int): Step to roll back to.
        """

        ptr_s = self.index_of_step(s)
        self.rollback_step_with_index(ptr_s)

        while self.latest() != -1 and self.latest() >= s:
            #print(f"rollback calling drop_last s={s} latest={self.latest()}")
            self.steps.drop_last();
			
        #self.steps.drop_last(self.steps.size() - ptr_s)
        #print('rollback!')
        #print( self.current )

    def estimate(self, s=None):
        """
        Obtain the most recent estimate of the state of a step.

        Args:
            s (int, optional): Step index. Defaults to latest step.

        Returns:
            tuple: (estimate, covariance matrix).
        """
        if self.steps.size() == 0:
            raise ValueError("No steps available for estimation.")

        if s is None:
            s = self.latest()
            
        ptr_s = self.index_of_step(s)
        #print(f"estimate step={s} ptr_s={ptr_s} earliest={self.earliest()} latest={self.latest()}")
        #print(self.steps.data.keys())
        step_data = self.steps.data[ptr_s]

        if "estimatedState" in step_data and step_data["estimatedState"] is not None:
            return step_data["estimatedState"], step_data["estimatedCovariance"]
        else:
            #import numpy as np
            #dim = step_data["dimension"]
            #return np.full((dim, 1), np.nan), np.full((dim, dim), np.nan)
            dim = step_data["dimension"]
            estimate = np.full((dim, 1), np.nan)
            # Return a CovarianceMatrix object instead of a raw NumPy array
            cov = CovarianceMatrix(np.eye(dim) * np.nan, 'C')
            #print("Warning: state is currently underdetermined")
            return estimate, cov            

    def gather(self):
        """
        Return all estimates concatenated into a single column vector.

        Returns:
            np.ndarray: Concatenated estimates.
        """
        import numpy as np

        N = sum(step["dimension"] for step in self.steps.data.values())
        u = np.zeros((N, 1))

        j = 0
        for step in self.steps.data.values():
            dim = step["dimension"]
            u[j:j + dim] = self.estimate(step["step"])[0]  # Extract state estimate
            j += dim

        return u
